### PR TITLE
Hide new type-based highlighting behind flag g:ocaml_use_v2_syntax

### PR DESCRIPTION
**Pull request by copy**
cc @c-cube @Maelan

### CODE REVIEW

**Comment by Maelan** [comment]
I have merely looked at the patch and haven’t tried it, but does it works as expected when the flag is *on*? Order of declarations in this file matters, so the large block that corresponds to the optional type linter should probably stay at the same place in the file. Also, the syntax elements that are added back by this patch should probably be included only when the flag is *off*, or they will conflict with the now-optional type linter.

File: syntax/ocaml.vim [hunk 1]
```diff
@@ -134,7 +775,7 @@ syn match    ocamlKeyword "\<include\>" skipwhite skipempty nextgroup=ocamlModPa
 " "module" - somewhat complicated stuff ;-)
 syn region   ocamlModule matchgroup=ocamlKeyword start="\<module\>" matchgroup=ocamlModule end="\<\u\(\w\|'\)*\>" contains=@ocamlAllErrs,ocamlComment skipwhite skipempty nextgroup=ocamlPreDef
 syn region   ocamlPreDef start="."me=e-1 matchgroup=ocamlKeyword end="\l\|=\|)"me=e-1 contained contains=@ocamlAllErrs,ocamlComment,ocamlModParam,ocamlGenMod,ocamlModTypeRestr,ocamlModTRWith nextgroup=ocamlModPreRHS
-syn region   ocamlModParam start="(\*\@!" end=")" contained contains=ocamlGenMod,ocamlModParam,ocamlModParam1,ocamlSig,ocamlVal
```
Comment (Maelan) [inline syntax/ocaml.vim:137]: This is reverting an unrelated change (which I believe was a bugfix).

File: syntax/ocaml.vim [hunk 2]
```diff
@@ -148,7 +789,7 @@ syn region   ocamlVal matchgroup=ocamlKeyword start="\<val\>" matchgroup=ocamlLC
 syn region   ocamlModRHS start="." end=". *\w\|([^*]"me=e-2 contained contains=ocamlComment skipwhite skipempty nextgroup=ocamlModParam,ocamlFullMod
 syn match    ocamlFullMod "\<\u\(\w\|'\)*\( *\. *\u\(\w\|'\)*\)*" contained skipwhite skipempty nextgroup=ocamlFuncWith
 
-syn region   ocamlFuncWith start="([*)]\@!" end=")" contained contains=ocamlComment,ocamlWith,ocamlFuncStruct skipwhite skipempty nextgroup=ocamlFuncWith
```
Comment (Maelan) [inline syntax/ocaml.vim:151]: ditto

File: syntax/ocaml.vim [hunk 3]
```diff
@@ -160,34 +801,24 @@ syn region   ocamlWithRest start="[^)]" end=")"me=e-1 contained contains=ALLBUT,
 syn region   ocamlStruct matchgroup=ocamlStructEncl start="\<\(module\s\+\)\=struct\>" matchgroup=ocamlStructEncl end="\<end\>" contains=ALLBUT,@ocamlContained,ocamlEndErr
 
 " "module type"
-syn region   ocamlKeyword start="\<module\s\+type\(\s\+of\)\=\>" matchgroup=ocamlModule end="\<\w\(\w\|'\)*\>" contains=ocamlComment skipwhite skipempty nextgroup=ocamlMTDef
```
Comment (Maelan) [inline syntax/ocaml.vim:163]: ditto

**Comment by Maelan** [comment]
Ah my bad, I did not see that the big `if` was ending with a `finish`. So you actually have 2 versions of the syntax with totally distinct implementations. But that doesn't sound maintainable.

**Comment by copy** [comment]
> So you actually have 2 versions of the syntax with totally distinct implementations. But that doesn't sound maintainable.

I agree, but I found this option more appealing than the other options:

- Reverting the changes and leaving them in a v2 branch: It would probably bitrot, even though the new highlighting is quite nice and works for some people (at least me)
- Doing nothing: In some colour schemes, the error highlighting looks pretty bad, so it probably shouldn't be on by default

That said, perhaps with the latest fixes we're the highlighting doesn't break as bad anymore.

**Comment by copy** [comment]
Let's close this, the new highlighting works pretty well.
